### PR TITLE
fused CUDA kernel for puzzlesim

### PR DESCRIPTION
**Pull request by Arcanous98**
This PR adds new CUDA kernels to accelerate Puzzlesim

* A set of custom CUDA kernels to accelerate key Puzzlesim operations (cosine similarity, bilinear upsampling) with arbitrary model support and reduced precission modes. Supports backward passes as well (~1.6x faster than the previous Pytorch backend)
* A fully fused CUDA kernel, limited for now to the Squeezenet model backbone. Supports forward evaluation only (~3x faster than the Pytorch backend)

In the future, we could extend the fused kernel to support arbitrary architecture backbones and derivatives to use Puzzlesim as a loss function using this backend.

### CODE REVIEW

**Comment by nihermann** [comment]
this pr will close #4.
